### PR TITLE
add unshare apparmor profile test

### PR DESCRIPTION
**Pull request by jessfraz**
ping @ewindisch @NathanMcCauley @diogomonica

### CODE REVIEW

File: integration-cli/docker_cli_run_test.go [hunk 1]
```diff
@@ -3174,3 +3174,14 @@ func (s *DockerSuite) TestMountIntoSys(c *check.C) {
 		c.Fatal("container should be able to mount into /sys/fs/cgroup")
 	}
 }
+
+func TestRunUnshareProc(c *check.C) {
+	testRequires(c, Apparmor)
+	defer deleteAllContainers()
+
+	name := "acidburn"
+	runCmd := exec.Command(dockerBinary, "run", "--name", name, "jess/unshare", "sh", "-c", "unshare -p -m -f -r --mount-proc=/proc mount")
```
Comment (tianon) [inline integration-cli/docker_cli_run_test.go]: Why `sh -c` and not just hit `unshare` directly?

Comment (tianon) [inline integration-cli/docker_cli_run_test.go]: ie:
```go
runCmd := exec.Command(dockerBinary, "run", "--name", name, "jess/unshare", "unshare", "-p", "-m", "-f", "-r", "--mount-proc=/proc", "mount")
```

Comment (jfrazelle) [inline integration-cli/docker_cli_run_test.go]: I want to say that didn't work but I can't remember so I'll try again haha

Comment (ewindisch) [inline integration-cli/docker_cli_run_test.go]: For some reason I've found that dropping *--mount-proc=/proc* and using *mount -t proc none /proc* works across more systems. I'm not sure if that's an issue with the versions of unshare that I've used, or the kernels.

Comment (jfrazelle) [inline integration-cli/docker_cli_run_test.go]: I can add a command to try both ways

Comment (jfrazelle) [inline integration-cli/docker_cli_run_test.go]: ya doesn't work withuout sh -c

**Comment by jfrazelle** [comment]
Updated PTAL

**Comment by tianon** [comment]
LGTM (nice container names :+1:)

**Comment by jfrazelle** [comment]
Also I solemly swear to keep this image up to date and beautiful :angel:

**Comment by ewindisch** [comment]
LGTM - although I think the '--mount-proc' version of the test is now redundant.

**Comment by jfrazelle** [comment]
but but but acidburn AND cereal ;)

**Comment by jfrazelle** [comment]
shit  I need to upgrade the PR servers

**Comment by jfrazelle** [comment]
seeing some weirdness in that the integration-daemon is not adding the process to enforce mode even tho the profile is loaded... so stay tuned

**Comment by jfrazelle** [comment]
ok got it... it's a bit hacky to say the least

**Comment by jfrazelle** [comment]
but at least now we will have a way to test the profile in the future ;)

**Comment by jfrazelle** [comment]
update PTAL

@tianon don't kill me :innocent:

File: hack/make/.integration-daemon-start [hunk 2]
```diff
@@ -27,6 +27,17 @@ if [ -n "$DOCKER_STORAGE_OPTS" ]; then
...
 
 if [ -z "$DOCKER_TEST_HOST" ]; then
+	# Start apparmor if it is enabled
+	if [ "$(cat /sys/module/apparmor/parameters/enabled)" == "Y" ]; then
+		# reset container variable so apparmor profile is applied to process
+		# see https://github.com/docker/libcontainer/blob/master/apparmor/apparmor.go#L16
+		export container=""
+		(
+			set -x;
+			exec /etc/init.d/apparmor start
```
Comment (tianon) [inline hack/make/.integration-daemon-start]: Why the semicolon and why the `exec` ?

**Comment by tianon** [comment]
Isn't apparmor a kernel feature?  What does the init script actually enable?

**Comment by jfrazelle** [comment]
it needs to be started to add the profile to the process

**Comment by jfrazelle** [comment]
and  ya idk but does not work without starting it _eyeroll_

File: hack/make/.integration-daemon-start (hunk 2 above)
Comment (jfrazelle) [inline hack/make/.integration-daemon-start]: copy paste eeeee

**Comment by tianon** [comment]
Totally insane seeming, but LGTM :+1:

**Comment by tianon** [comment]
The way it's coded now, it'll only run in the build container, since folks running these tests outside that need to be using `DOCKER_TEST_HOST`.

**Comment by ewindisch** [comment]
@tianon I see, DOCKER_TEST_HOST is set if you want to run tests outside of a container? In this case, it will still be modifying apparmor config on the host. We should at least warn about about that somewhere.

**Comment by jfrazelle** [comment]
it only does this if `DOCKER_TEST_HOST` is empty, which would imply its not running directly on some remote host

**Comment by jfrazelle** [comment]
I swear this is safe you guys

(famous last words) but trust me

File: integration-cli/docker_cli_run_test.go [hunk 3]
```diff
@@ -3211,3 +3211,19 @@ func (s *DockerSuite) TestTwoContainersInNetHost(c *check.C) {
...
+	testRequires(c, Apparmor)
+
+	name := "acidburn"
+	runCmd := exec.Command(dockerBinary, "run", "--name", name, "jess/unshare", "unshare", "-p", "-m", "-f", "-r", "--mount-proc=/proc", "mount")
+	if out, _, err := runCommandWithOutput(runCmd); err == nil || !strings.Contains(out, "Permission denied") {
+		c.Fatalf("unshare should have failed with permission denied, got: %s, %v", out, err)
+	}
+
+	name = "cereal"
+	runCmd = exec.Command(dockerBinary, "run", "--name", name, "jess/unshare", "unshare", "-p", "-m", "-f", "-r", "mount", "-t", "proc", "none", "/proc")
```
Comment (NathanMcCauley) [inline integration-cli/docker_cli_run_test.go]: +1 I <3 tests

**Comment by NathanMcCauley** [comment]
LGTM

**Comment by jfrazelle** [comment]
who will click the button ;)

**Comment by thaJeztah** [comment]
> who will click the button ;)

It's tempting; it's nice and shiny and green!

**Comment by ewindisch** [comment]
LGTM

**Comment by thaJeztah** [comment]
Asked @ewindisch  if it's safe to merge, just to be sure. And his "famous last words" were "it should be fine"

So, here we go!

**Pull request merged**